[doc] Add user utilities doc (#1295)
taichi-dev/taichi: docs/utilities.rst

Developer utilities
===================

Taichi provides some tools to help you develop it efficiently. This section tries to provide a detailed description of some commonly used utilities.

Logging
-------

Taichi provides logging APIs based on ``spdlog``. Note that these functions will only be **called in compile-time once** but not run-time.

If you want to do logging in **run-time**, please use ``print()`` instead.

.. Note::

    Taichi logging APIs only support standard output now.

.. function:: ti.set_logging_level(level)

    :parameter level: (string) a valid logging level

    - This function is used to set the level of logging. Currently the log levels in Taichi are sorted as ``ti.TRACE``, ``ti.DEBUG``, ``ti.INFO``, ``ti.WARN`` and ``ti.ERROR`` in descending order. The default logging level is ``ti.INFO``.

    - These logging levels are essentially strings ``trace``, ``debug``, ``info``, ``warn`` and ``error`` respectively.

> 

Author: Okay :)

.. code-block:: python

    import taichi as ti
    ti.init(arch=ti.cpu)

    ti.set_logging_level(ti.INFO)
    var = ti.var(ti.f32, shape=1)

    @ti.kernel
    def compute():
        var[0] = 1.0
        ti.info(f"set var[0] = 1.0")

    compute()

.. function:: ti.warn(info)

    :parameter info: (string) logging info

    Print the input string in yellow color to stdout **only in Taichi-scope**, when the logging level is lower or equal to ``ti.WARN``.

.. function:: ti.debug(info)

    :parameter info: (string) logging info

    Print the input string in blue color to stdout **only in Taichi-scope**, when the logging level is lower or equal to ``ti.DEBUG``.

.. function:: ti.trace(info)

    :parameter info: (string) logging info

    Print the input string in green color to stdout **only in Taichi-scope**, only when the logging level is equal to ``ti.TRACE``.

.. function:: ti.error(info)

    :parameter info: (string) logging info

    This function prints the input string in red color to stdout in any logging level and **only in Taichi-scope**.

.. warning::

>     Note that ``ti.error`` will crash your program and throws an exception ``RuntimeError``.

Author: Good idea! Please let me simplify this part and submit a new version.

Here is an example:

.. code-block:: python

    import taichi as ti

    ti.init()
    ti.set_logging_level(ti.INFO)

    try:
        ti.error("It's a big error! Now exit...")
    except RuntimeError as err:
        print(err)

Profiler
--------

Taichi profiler can help you count the time cost of your program. There are two profilers in Taichi available at this moment: ``ScopedProfiler`` and ``ProfilerBase``.

ScopedProfiler
##############

> 1. ``ScopedProfiler`` evaluates the time spent for each step of the **host tasks** during execution hierarchically.

Reviewer: ```suggestion
1. ``ScopedProfiler`` measures time spent on the **host tasks** hierarchically.
```

> 2. Moreover, this profiler is turned on automatically and no extra work is needed except writing ``ti.print_profile_info()`` at the end of your Taichi program. For example:

Reviewer: ```suggestion
2. This profiler is turned on automatically. To show its results, call ``ti.print_profile_info()``. For example:
```

.. code-block:: python

    import taichi as ti

    ti.init(arch=ti.cpu)
    var = ti.var(ti.f32, shape=1)

    @ti.kernel
    def compute():
        var[0] = 1.0

>         print(f" set var[0] = {var[0]}")

Reviewer: ```suggestion
        print(f"set var[0] =", var[0])
```

    compute()
    ti.print_profile_info()

> ``ti.print_profile_info()`` will output statistics in a hierarchical format, with different colors for different levels.

Reviewer: ```suggestion
``ti.print_profile_info()`` prints profiling results in a hierarchical format.
```

.. Note::

>     ``ScopedProfiler`` is a c++ class inside the core of Taichi, so it is invisible in Python-scope.We mention its name here is just to make a distinction with another profiler.

Reviewer: ```suggestion
    ``ScopedProfiler`` is a C++ class in the core of Taichi. It is not exposed to Python users.
```

> ProfilerBase

Reviewer: This should be called `KernelProfiler`.

Author: Oh! The old `ProfilerBase` has been renamed to `KernelProfilerBase`, I didn't manage to notice it...

############

> 1. ``ProfilerBase`` records the cost of kernel launches on devices. ``kernel_profiler`` must be set to ``True`` in ``ti.init`` to enable this profiler.

Reviewer: ```suggestion
1. ``KernelProfiler`` records the costs of Taichi kernels on devices. To enable this profiler, set ``kernel_profiler=True`` in ``ti.init``.
```

> 2. When all the computations are finished in your program, call ``ti.kernel_profiler_print()`` to show the profiling result. For example:

Reviewer: ```suggestion
2. Call ``ti.kernel_profiler_print()`` to show the kernel profiling result. For example:
```

No need to wait for all computations are finished.

.. code-block:: python
    :emphasize-lines: 3, 11

    import taichi as ti

>     ti.init(ti.cpu, kernel_profiler = True)

Reviewer: ```suggestion
    ti.init(ti.cpu, kernel_profiler=True)
```

>     var = ti.var(ti.f32, shape = 1)

Reviewer: ```suggestion
    var = ti.var(ti.f32, shape=1)
```

Author: I will format the example code even in the doc next time 👍 :(

    @ti.kernel
    def compute():
        var[0] = 1.0

    compute()
    ti.kernel_profiler_print()

> Its output is like:

Reviewer: ```suggestion
The outputs would be:
```

Developer utilities
===================

This section provides a detailed description of some commonly used utilities for Taichi developers.

> Logging

Reviewer 2: Hi, sorry about dismissing your changes, I have already done with `Logging` in #1475 :)

Author: May you merge these 2 changes into a single one? I also have no idea about how to deal with this conflict.

Reviewer 2: To prevent conflicts with #1475, you may move the `Profiler` to another section like `profiler.rst` :) I'll try to merge conflicts in `Logging` section once one of these two PR is merged.

::

    [I 07/09/20 13:15:24.517] [main.py:calc@10] set var in ti.func
    [I 07/09/20 13:15:24.518] [main.py:calc@10] set var in ti.func
    [I 07/09/20 13:15:24.518] [main.py:calc@10] set var in ti.func
    [I 07/09/20 13:15:24.518] [main.py:compute@19] set var in ti.kernel

The other logging functions below **are all similar to** ``ti.info``. They can all print out when the logging level is set to be lower than they required, respectively.

.. function:: ti.warn(info)

    :parameter info: (string) logging info

.. function:: ti.debug(info)

    :parameter info: (string) logging info

.. function:: ti.trace(info)

    :parameter info: (string) logging info

.. function:: ti.error(info)

    :parameter info: (string) logging info

    This function prints the input string in any logging level and **crashes the program**.

.. warning::

    Note that ``ti.error`` will crash your program and throws an exception ``RuntimeError``.

Here is an example:

.. code-block:: python

    import taichi as ti

    ti.init()
    ti.set_logging_level(ti.INFO)

    try:
        ti.error("Fatal error. Exiting now...")
    except RuntimeError as err:
        print(err)

> Profiler

Reviewer 2: Profiler should be shared by devs and end-users, maybe we can have a separate .rst for it?

Author: IMO I think it's enough in `utilities.rst` at this moment.
